### PR TITLE
Raw json support

### LINKED ISSUE #143: indices client and xcontent/json support
Hello and I have a question:
I am trying out this new fluent client and  have a bunch of projects that I wanted to migrate from the RHLC to the new API client. Now I am kinda stuck on an inconvenient spot, creating an index with settings and mapping. the builder and functional pattern are nice, though in our projects we put json-files for mappings and settings which are read and converted into an object/xContentBuilder so we do not have to implement complex settings/mappings into java code (and reuse without copy paste).

kinda like that
```
CreateIndexRequest createIndexRequest = new CreateIndexRequest(indexName)
				.settings(this.elasticsearchIndexService.createSettings(ELASTICSEARCH_INDEX_SETTINGS_FILE))
				.mapping(this.elasticsearchIndexService.createMapping(ELASTICSEARCH_INDEX_MAPPING_FILE,
					ELASTICSEARCH_INDEX_MAPPING_FOLDER, shop));
AcknowledgedResponse indexCreateResponse = this.restHighLevelClient.indices().create(createIndexRequest, RequestOptions.DEFAULT);
```
depending on the `shop` parameter different mappings are loaded.

Is there a similar and fluent way doing that in the new client or are there any plans supporting json/xcontent?

### PR DESCRIPTION
**Pull request by swallez**
Fixes #143

Adds `withJson` methods to object builders. These methods populate an object builder from arbitrary JSON by using its deserializer. This allows creating API objects with a combination of programmatic call to property setters and arbitrary JSON data.

The main use cases are loading search requests, index mappings or documents from arbitrary JSON sources.

### CODE REVIEW

**Comment by swallez** [comment]
@spinscale @dadoonet can you have a look at [the docs](https://elasticsearch-java_200.docs-preview.app.elstc.co/guide/en/elasticsearch/client/java-api-client/master/loading-json.html) and [the `WithJson` API](https://github.com/elastic/elasticsearch-java/pull/200/files#diff-fc7fc3c8f22b4078b6971379500b3efd8dcd9ea5bfde9d1959a6ba86be8a8e65) for this feature and tell me what you think?

@szabosteve can you do a quick review of [the docs](https://github.com/elastic/elasticsearch-java/pull/200/files#diff-bdf87bf02df505fc9b0dadbed898ed7393efb76d5b1745f80781c1d76c348f60) ([rendered](https://elasticsearch-java_200.docs-preview.app.elstc.co/guide/en/elasticsearch/client/java-api-client/master/loading-json.html))?

**Comment by dadoonet** [comment]
I'm wondering why do we need to provide the `jsonpMapper`:

```java
CreateIndexRequest req = CreateIndexRequest.of(b -> b
    .index("some-index")
    .withJson(
        jsonStream, 
        client._jsonpMapper() 
    )
);
```

Would it be possible to have a default one like this:

```java
CreateIndexRequest req = CreateIndexRequest.of(b -> b
    .index("some-index")
    .withJson(jsonStream)
);
```

**Comment by dadoonet** [comment]
About the documentation, I think that each example should be self contained. So for [the last section](https://elasticsearch-java_200.docs-preview.app.elstc.co/guide/en/elasticsearch/client/java-api-client/master/loading-json.html#_creating_a_search_request_from_multiple_json_snippets), add again the `queryJson`:

```java
Reader queryJson = new StringReader(
    "{" +
    "  \"query\": {" +
    "    \"range\": {" +
    "      \"@timestamp\": {" +
    "        \"gt\": \"now-1w\"" +
    "      }" +
    "    }" +
    "  }" +
    "}");
```

File: java-client/src/test/java/co/elastic/clients/documentation/LoadingJson.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            .withJson(file, client._jsonpMapper())
+        );
+
+        client.index(req);
+        //end::ingest-data
+    }
+
+
+    public void query1() throws IOException {
+        //tag::query
```
Comment (dadoonet) [inline java-client/src/test/java/co/elastic/clients/documentation/LoadingJson.java:94]: I'd write:

```java
        {
        //tag::query
        Reader queryJson = new StringReader(
            "{" +
            "  \"query\": {" +
            "    \"range\": {" +
            "      \"@timestamp\": {" +
            "        \"gt\": \"now-1w\"" +
            "      }" +
            "    }" +
            "  }" +
            "}");

        SearchRequest agg1 = SearchRequest.of(b -> b
            .withJson(queryJson, client._jsonpMapper()) //<1>
            .aggregations("max-cpu", a1 -> a1 //<2>
                .dateHistogram(h -> h
                    .field("@timestamp")
                    .calendarInterval(CalendarInterval.Hour)
                )
                .aggregations("max", a2 -> a2
                    .max(m -> m.field("host.cpu.usage"))
                )
            )
            .size(0)
        );

        Map<String, Aggregate> aggs1 = client
            .search(agg1, Void.class) //<3>
            .aggregations();
        //end::query
        }

        {
        //tag::query-and-agg
        Reader queryJson = new StringReader(
            "{" +
            "  \"query\": {" +
            "    \"range\": {" +
            "      \"@timestamp\": {" +
            "        \"gt\": \"now-1w\"" +
            "      }" +
            "    }" +
            "  }" +
            "}");

        Reader aggJson = new StringReader(
            "\"size\": 0, " +
            "\"aggs\": {" +
            "  \"hours\": {" +
            "    \"date_histogram\": {" +
            "      \"field\": \"@timestamp\"," +
            "      \"interval\": \"hour\"" +
            "    }," +
            "    \"aggs\": {" +
            "      \"max-cpu\": {" +
            "        \"max\": {" +
            "          \"field\": \"host.cpu.usage\"" +
            "        }" +
            "      }" +
            "    }" +
            "  }" +
            "}");

        SearchRequest agg2 = SearchRequest.of(b -> b
            .withJson(queryJson, client._jsonpMapper()) //<1>
            .withJson(aggJson, client._jsonpMapper()) //<2>
            .ignoreUnavailable(true) //<3>
        );

        Map<String, Aggregate> aggs2 = client
            .search(agg2, Void.class)
            .aggregations();
        //end::query-and-agg
        }
```

File: java-client/src/main/java/co/elastic/clients/util/WithJsonObjectBuilderBase.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ * <p>
+ * If the concrete class has generic parameters, <code>withJson()</code> will deserialize them as {@link JsonData}, unless
+ * the mapper has a deserializer attribute for that generic parameter's name.
+ */
+public abstract class WithJsonObjectBuilderBase<B> extends ObjectBuilderBase implements WithJson<B> {
+
+    protected abstract B self();
+
+    @Override
+    public B withJson(JsonParser parser, JsonpMapper mapper) {
```
Comment (dadoonet) [inline java-client/src/main/java/co/elastic/clients/util/WithJsonObjectBuilderBase.java:43]: Would it be possible to have a default json mapper that we can use as well?
So the method would be:

```java
public B withJson(JsonParser parser)
```

**Review by dadoonet — COMMENTED** [review]
Left some comments.

File: java-client/src/main/java/co/elastic/clients/util/WithJsonObjectBuilderBase.java (hunk 2 above)
Comment (swallez) [reply java-client/src/main/java/co/elastic/clients/util/WithJsonObjectBuilderBase.java:43]: Addressed in 78e8ee2

**Comment by spinscale** [comment]
So I pretty much only looked at `LodingJsonTest.java`. Few minor comments.

First, can one input overwrite another? The samples only show things along side. What happens in `query2()` if both set `size`? I suppose it gets overwritten?

Second, still thinking about the wording of `withJson`. With the additivity that probably makes more sense than `addJson/ofJson` or similar.

Third, what happens on invalid JSON (or duplicate keys)? Early error when trying to parse JSON or is a wrong request sent to Elasticsearch?

The `loadIndexDefinition()` is a great example, as I expect this to be the number one source of things.

The samples would be a great candidate for text blocks, but I suppose, we need to retain Java BWC here :-)

**Comment by swallez** [comment]
@spinscale answers below

> First, can one input overwrite another? The samples only show things along side. What happens in query2() if both set size? I suppose it gets overwritten?

That's right. You can also overwrite properties loaded from JSON programatically. Key point is that it overwrites existing values and doesn't merge them (this would involve complex rules, both to implement and understand). I will add a section about it in the docs to make it clear and unambiguous.

> Second, still thinking about the wording of withJson. With the additivity that probably makes more sense than addJson/ofJson or similar.

Ok, so we keep `withJson` then 😉 

> Third, what happens on invalid JSON (or duplicate keys)? Early error when trying to parse JSON or is a wrong request sent to Elasticsearch?

This goes through the object's deserializer ("partial deserialization" as said in the `WithJson` interface javadoc), so errors are raised on unknown fields or invalid value types. There's no way to send data to Elasticsearch that could not have been built programmatically. 

> The samples would be a great candidate for text blocks, but I suppose, we need to retain Java BWC here :-)

Uh, I don't understand what you mean. Some doc system feature I'm not aware of? Can you elaborate?

**Review by spinscale — APPROVED** [review]

**Comment by swallez** [comment]
> Regarding your third point: I need to think if this is a good idea or not, as this method offers a nice way out of the problem to provide java client extensions for custom plugin functionality. I'd start with your approach and check for feedback?

Custom plugins that add new API endpoints and/or new component types for queries, aggregations, etc are a different topic that will be addressed specifically. Typically for components we should have a `custom` variant that accepts any JSON structure or object that can be serialized/deserialized.

**Comment by jsnyder4** [comment]
Can we get a new release with enhancement? (perhaps 8.1.1 or 8.2.0) I'm new to elasticsearch, so I was reading the [official documentation for 8.1](https://www.elastic.co/guide/en/elasticsearch/client/java-api-client/current/loading-json.html) and saw `.withJson()`. To my surprise, the method is not actually available in 8.1.0, which I didn't realize until finally stumbling on this pull request.

**Comment by swallez** [comment]
@jsnyder4 it is part of 8.1.1 that was released on March 18th.

**Comment by jsnyder4** [comment]
Yeah, I see that now. Thank you very much.

**Comment by jepson66** [comment]
hi,swallez
I have two questions:
1. Using withjson, the original json format will be changed, for example, the original order is script_fields, query, build SearchRequest.of(b -> b
                 After .withJson(new StringReader(reqJson)).size(0)), it becomes the order of query and script_fields, so I can't find the data, the original json is executed normally in kinaba
2. In springboot2.7.1 using elasticsearch of version 8.2.2, RestHighLevelClient cannot be used. According to [official documents](https://www.elastic.co/guide/en/elasticsearch/client/java-api-client/8.2/migrate-hlrc.html) and test cases, start reporting version conflict

**Pull request merged**